### ISSUE TITLE
Reproducibility audit: HR-Agent twin-env training is dead code; broken imports; missing feature-extraction pipeline

### ISSUE THREAD

**Issue by KYUYULLEE-SQR** [Edwicn/OPHR-MasteringVolatilityTradingwithMultiAgentDeepReinforcementLearning#3]
## Reproducibility audit — several blocking issues from a code-trace

Hi authors — congrats on the NeurIPS 2025 acceptance. We are independently auditing the released code with the goal of comparing OPHR against our own crypto-options RL setup. After a careful read of the repo (commit at HEAD as of 2026-05-01) we ran into several issues that block reproduction of the paper's headline claims. Filing them here in case you want to confirm/clarify or push a patch.

The audit was code-only (we did not have to invoke any private data). All findings reference files at the paths listed.

---

### 1. `training/__init__.py` imports a module that does not exist

```
training/__init__.py  →  from .phase2_twin_env import ...
```
There is no `phase2_twin_env.py` in `training/`. Any import of the `training` package fails immediately, so neither `phase1_oracle.py`, `phase1_op_offline.py`, `phase1_hr_warmup.py`, nor `phase2_iterative.py` can be invoked through normal package imports. We had to bypass `__init__.py` to even reach a smoke run.

### 2. `TrainingConfig.from_yaml` reads fields not present in shipped YAML

`config.py` `TrainingConfig.from_yaml` reads several keys (e.g. fields under the `op_agent` and `hr_agent` blocks, plus a few top-level training fields) that are absent from `configs/training_config.yaml` as shipped. This raises a `KeyError` on the very first config load. We had to patch in default values (we are happy to share the patch list if useful) before the rest of the pipeline could be exercised.

### 3. **HR-Agent twin-env training is dead code in both `phase1_hr_warmup.py` and `phase2_iterative.py`**

This is the one we are most concerned about, because the twin-env relative-reward update for HR-Agent is, as we read the paper, the headline algorithmic contribution.

Evidence:

- `agents/hr_agent.py:223–240` — `should_make_decision()` returns `True` only when `steps_since_decision >= n_hr`. `hr_agent.step(...)` itself contains an internal `if self.should_make_decision(): ... reset steps_since_decision = 0` block.
- `training/phase2_iterative.py:152` — the entire twin-env block (state save → temp rollout → restore → twin rollout → relative-reward computation → `hr_agent.update()`) is gated by an **outer** `if self.hr_agent.should_make_decision():`. The else-branch at line 249 calls `hr_agent.step(...)`, which always resets `steps_since_decision` before it can reach `n_hr`.
- `training/phase1_hr_warmup.py:226 + 348` — same pattern. The outer `if hr_agent.should_make_decision():` (line 226) wraps the warmup twin-env block; the else-branch at line 348 calls `hr_agent.step(...)` and self-resets the trigger.

In both phases the outer condition can never fire, so the twin-env relative-reward training loop is never entered. Empirically confirmed in our smoke run: `train_hr_episode` reported `hr_decisions=0` over a 1920-step rollout, and HR-Agent therefore ends training with its random initial weights — `hr_agent.update()` (line 313 in warmup, line 229 in iterative) is never reached.

If we are reading this correctly, Algorithm 2 in the paper (alternating OP/HR with twin-env relative reward) is not implemented in the released code. If the paper's results are nonetheless correct, it would be very helpful to release the actual training script that produced them — the current `phase2_iterative.py` does not.

### 4. `perp.parquet` requires 905 anonymized engineered features (no extraction script)

The shipped `sample_data/BTC/perp.parquet` schema includes columns `feature_0..feature_904`. We could not find any script in the repo that produces these from raw Deribit data, and the README is empty. Could you publish the feature-extraction pipeline (or document the columns)? Without this it is not possible to retrain on different time periods (e.g. for stress tests on the Oct-2025 BTC drawdown). The 25-level orderbook depth columns also seem to require a different source than Deribit's public feed.

### 5. Minor — paper vs. code inconsistencies

- `configs/env_config.yaml` sets `fee_config.futures_perpetual: 0.0000`. The paper text states 0.05% perpetual fees. Even if Table 4's reported tcost ratios are correct in absolute USD, the relative split between options vs. perp costs depends on this field, and a 0% perp fee in training will systematically under-hedge.
- `valid_option_dict.pkl` is a 16 MB binary blob with no schema documentation. A JSON or parquet replacement (or at least a one-paragraph format spec) would help downstream users.
- DQN loss diverges to ~1e28 within ~5 batches in our smoke run (no gradient clipping, no Huber loss, no reward normalisation). Probably not user-facing in the originals' setup, but it makes the released training scripts hard to run from scratch.

### 6. Concern about Oracle distillation indirect leak

Oracle policy uses `lookforward_window: 168h` (1 week) RV-vs-IV labels for OP-Agent BC pretraining. The paper rightly notes the labels are "used solely to generate" the distillation target. But a head-to-head ablation (Phase II from scratch, no Oracle pretrain) would be reassuring and is missing from the current Table 2 (the OP ablation listed there appears to still inherit the Oracle-distilled init). Could you confirm whether the OP-only baseline includes Oracle pretraining or not?

---

We genuinely like the architectural direction (decoupled OP/HR with env-state communication, multi-hedger routing). But as currently released, points 1–4 prevent independent reproduction of the headline numbers, and point 3 calls into question whether the paper's results were generated by the algorithm as described in the code.

Happy to send patches / share our code trace if helpful. Thanks for your time.
